src/gradient.jl

```diff
@@ -80,13 +80,13 @@ function extract_gradient!(::Type{T}, result::DiffResult, dual::Dual) where {T}
 end
 
 extract_gradient!(::Type{T}, result::AbstractArray, y::Real) where {T} = fill!(result, zero(y))
-extract_gradient!(::Type{T}, result::AbstractArray, dual::Dual) where {T}= copyto!(result, partials(T, dual))
+function extract_gradient!(::Type{T}, result::AbstractArray, dual::Dual) where {T}
+    result[:] .= partials(T, dual)
```

Comment:
Why the the `[:]`?
```
julia> @btime $(rand(3,3))[:] .= 0;
  min 27.973 ns, mean 35.812 ns (2 allocations, 80 bytes)

julia> @btime $(rand(3,3)) .= 0;
  min 4.541 ns, mean 4.649 ns (0 allocations)
```

Reply:
Thats for the case where `result::Matrix` since `partials::AbstractVector`, otherwise its a size-mismatch error for those cases. That's exactly the allocations I was talking about above.

Reply:
Oh I see. Will it work if it's `Base.ReshapedArray((rand(3,3)), (9,), ())  .= 0`?

Reply:
Or maybe
```suggestion
    vec(result) .= partials(T, dual)
```
or
```suggestion
    copyto!(result, partials(T, dual))
```
?

Reply:
Or
```suggestion
    fill!(result, partials(T, dual))
```
?

Reply:
`vec(result) .=` is the same as `results[:] .=`, and `fill!(rand(3), (1,2,3))` won't work.

Reply:
What about
```suggestion
    copyto!(result, partials(T, dual))
```
?

Reply:
That's exactly what it was before that doesn't work with CUDA :) 

I guess you could say its CUDA's fault they don't have `copyto!(::CuArray, ::AbstractArray)` or something, JLArrays doesn't seem to have it either fwiw. This generic version might enable more array types we're not thinking of right now. 

Here's three ideas: 

1. Keep PR as is and eat the allocation (only for `result::Matrix`, not `result::Vector`) in the name of simpler more generic code. 
2. Bring back old versions and use them for `::Array` and use my changes for `::AbstractArray`
3. Bring back old versions and use them for `::AbstractArray` and use my changes for `::AbstractGPUArray` (and then depend on that package)

Up to you guys.

Reply:
> I guess you could say its CUDA's fault they don't have copyto!(::CuArray, ::AbstractArray) or something, 

Yes, I would have assumed that CUDA defines `copyto!`, possibly via the broadcasting fallback. I think adding it to CUDA would be the optimal solution - it would not require any workarounds here and could be useful more generally.

Reply:
Sorry, this is all moot point, see my comment below.

Reply:
As an additional comment:

> Keep PR as is and eat the allocation (only for result::Matrix, not result::Vector) in the name of simpler more generic code.

IMO arguably `copyto!` is the simplest and most generic code. In contrast to the broadcasting "hack" it is concise and a function that is supposed to do exactly what we want to achieve here. It should work regardless of whether the array supports scalar indexing or not. (Also broadcasting tends to cause increased compilation times.)

Reply:
My problem is which packages' responsibility is `copyto!(::CuArray, ::Partials)` and if anything I'd argue its ForwardDiff's since Partials is the more limited-used array type (see also [this](https://github.com/JuliaGPU/GPUArrays.jl/issues/422#issuecomment-1207164226))

But maybe its a moot point, since I discovered this works without any allocations and for CuArrays: 
```julia
result[:] = partials(T, dual)
```

I'm still searching for something that works for the chunk mode though.

Reply:
> My problem is which packages' responsibility is copyto!(::CuArray, ::Partials) and if anything I'd argue its ForwardDiff's since Partials is the more limited-use array type (see also https://github.com/JuliaGPU/GPUArrays.jl/issues/422#issuecomment-1207164226)

IMO the natural thing to do would be to use `copyto!(result, partials(T, dual).values)` since in that case the second argument is just a `Tuple` which should work more generally.

Reply:
> is just a `Tuple` which should work more generally.

It's even a `NTuple`, so support for heterogeneous tuples is not needed.

Reply:
Thanks, yea, I do agree `copyto!` makes alot of logical sense for the operation we want, but I've got a solution below now which only relies on an even simper interface (basically get/set index). I can ask about the prospects of a `copyto!(::CuArray, ::Tuple)` though since it does seem to offer a good way to get stack-allocated memory onto GPU. If that ever happens we could then switch to `copyto!`.